docs: explain Kubernets resources used in Cyclone (#1337)
caicloud/cyclone: docs/develop/crd.md

## Kubernetes Original Resources

* **Namespace**: Cyclone is a Multi-Tenant Isolation system. 

    In general, there are two namespaces for each tenant:

    * Metadata namespace: Store Metadata for one Tenant, and hold the tenant description information in annotations. *MUST* in the Control cluster(cluster cyclone running in), named in format of `cyclone-{tenant-name}`

    * Workload namespace: Execute Workload. Maybe in the Control cluster OR other clusters. 

    However you can explicitly specify the two namespaces as one: `cyclone-{tenant-name}`, then Cyclone will store metadata and execute Workload in it(Not Recommend).

* **Secret**: Cyclone supports [integrate external systems](./user_guide.md#integration-center).Cyclone uses `Secret` to protect the auth information of external systems as it is pretty sensitive.

* **PVC**: Cyclone uses `PVC` to share data in different pods.

    There are two types of' PVC' Cyclone used:

    * Tenant Scope PVC: Each tenant has a `PVC` in the Workload namespace to 
        * Share runtime data among different Pods belongs to the same WorkflowRun.
        * (Optional) Cache data to accelerate Workflow's next time executing. 

>     * Cyclone-Server PVC: Usually we configure a `PVC` for Cyclone-Server to store and serve

Reviewer: I think there is no need to mention the PVC used by Cyclone server here. It is not required, users can use the disk of pod if they think the logs are not important .

Author: How about add a note:
```
If your Cyclone-Server has only one replica, the Cyclone-Server PVC is recommended instead of required, you can use the disk of pod if you think the logs and artifacts are not important. But if the replica of your Cyclone-Server is greater than one, the Cyclone-Server PVC is required as multiple Cyclone-Server instances need to share the data.
```

Reviewer: `Tenant Scope PVC` has special business logic, but `Cyclone server PVC` is just a general usage of PVC. I still prefer to not mentioning the PVC used by Cyclone server here, it should be part of Cyclone design doc.

        * Logs generated by Workflow Running
        * Artifacts generated by Workflow Running

> * **ResourceQuota**: Each tenant has a `ResourceQuota` in the Workload namespace to control CPU/Memory resources.

Reviewer: ```suggestion
* **ResourceQuota**: Each tenant has a `ResourceQuota` in the Workload namespace to limit CPU/Memory usage.
```

> * **Pod**: As we all known Pod are the smallest deployable units in Kubernetes and Cyclone is a Kubernetes native Workflow Engine, Cyclone creates and manages Pod to run tasks. 

Reviewer: ```suggestion
* **Pod**: As we all known Pod are the smallest schedulable units in Kubernetes, Cyclone creates and manages Pod to run tasks.
```

>     There are three types of' Pod' Cyclone used, and they all running in their own Workload namespace for each Tenant:

Reviewer: ```suggestion
    There are three types of pods used by Cyclone, and they all running in their own Workload namespace for each tenant:
```

>     * Workload pod(OR called Stage Pod): One stage a pod. A workload pod executes tasks defined in a stage.

Reviewer: ```suggestion
    * Workload pod: One pod per stage. A workload pod executes tasks defined in a stage.
```

>     * GC pod: One WorkflowRun a pod. As mentioned above that Tenant Scope PVC shares runtime data, and the runtime data needs to be cleaned after running completed. The GC pod does this job.

Reviewer: ```suggestion
    * GC pod: One pod per WorkflowRun. As mentioned above that Tenant Scope PVC shares runtime data, which is temporary data and needs to be cleaned after WorkflowRun completes. The GC pod does the cleaning work.
```

>     * PVC watchdog: One tenant a pod.  **Long-Running** in the Workload namespace. It watches the usages of Tenant Scope PVC and reports the usages information to Cyclone-Server periodically. 

Reviewer: ```suggestion
    * PVC watchdog: One pod per tenant. **Long-Running** pod in each Workload namespace. It watches the usages of Tenant Scope PVC and reports the usage information to Cyclone server periodically. 
```

# Kubernetes Resources Used by Cyclone

As we known Cyclone is implemented with native Kubernetes resources with no extra dependencies. Following we will illustrate all resources used by Cyclone, including Kubernetes originally defined and Cyclone defined.

## Kubernetes Original Resources

* **Namespace**: Cyclone is a multi-tenant isolated system.

    In general, there are two namespaces for each tenant:

    * Metadata namespace: stores metadata for one tenant, and holds its basic information in annotations. *MUST* be in the control cluster where Cyclone runs), named in format of `cyclone-{tenant-name}`

    * Workload namespace: Where workloads execute. Can be in control cluster OR user clusters.

>     However you can explicitly specify the two namespaces as one: `cyclone-{tenant-name}`, then Cyclone will store metadata and execute Workload in it(Not Recommend).

Author: Single responsibility, we can easily recover when workload namespace gets ruined if it is not responsible for storing metadata.
